src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc

```diff
@@ -0,0 +1,430 @@
+    return;
+  }
+
+  notify_quiesce();
+}
+
+template <typename I>
+void GroupImageCreatePrimaryRequest<I>::notify_quiesce() {
+  ldout(m_cct, 15) << dendl;
+
+  auto ctx = create_context_callback<
+    GroupImageCreatePrimaryRequest<I>,
+    &GroupImageCreatePrimaryRequest<I>::handle_notify_quiesce>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  m_quiesce_requests.resize(m_image_ctxs.size());
+
+  for (size_t i = 0; i < m_image_ctxs.size(); ++i) {
+    auto ictx = m_image_ctxs[i];
+    ictx->image_watcher->notify_quiesce(&(m_quiesce_requests)[i], m_prog_ctx,
+                                        gather_ctx->new_sub());
+  }
+
+  gather_ctx->activate();
+}
+
+template <typename I>
+void GroupImageCreatePrimaryRequest<I>::handle_notify_quiesce(int r) {
+  ldout(m_cct, 15) << "r=" << r << dendl;
+
+  if (r < 0 &&
+      (m_group_snap_create_flags & SNAP_CREATE_FLAG_IGNORE_NOTIFY_QUIESCE_ERROR) == 0) {
+    m_ret_code = r;
+    notify_unquiesce();
+    return;
+  }
+
+  acquire_exclusive_locks();
+}
+
+template <typename I>
+void GroupImageCreatePrimaryRequest<I>::acquire_exclusive_locks() {
+  ldout(m_cct, 15) << dendl;
+
+  m_release_locks = true;
+
+  auto ctx = librbd::util::create_context_callback<
+    GroupImageCreatePrimaryRequest<I>,
+    &GroupImageCreatePrimaryRequest<I>::handle_acquire_exclusive_locks>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  for (auto ictx: m_image_ctxs) {
+    std::shared_lock owner_lock{ictx->owner_lock};
+    if (ictx->exclusive_lock != nullptr) {
+      ictx->exclusive_lock->block_requests(-EBUSY);
+      ictx->exclusive_lock->acquire_lock(gather_ctx->new_sub());
+    }
+  }
+
+  gather_ctx->activate();
+}
+
+template <typename I>
+void GroupImageCreatePrimaryRequest<I>::handle_acquire_exclusive_locks(int r) {
+  ldout(m_cct, 15) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to acquire image exclusive locks: "
+                 << cpp_strerror(r) << dendl;
+    m_ret_code = r;
+    // release locks in case some of the lock acquisitions succeeded
+    release_exclusive_locks();
+    return;
+  }
+
+  create_snapshots();
+}
+
+template <typename I>
+void GroupImageCreatePrimaryRequest<I>::create_snapshots() {
+  ldout(m_cct, 15) << dendl;
+
+  auto ctx = create_context_callback<
+    GroupImageCreatePrimaryRequest<I>,
+    &GroupImageCreatePrimaryRequest<I>::handle_create_snapshots>(this);
+
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  for (size_t i = 0; i < m_image_ctxs.size(); i++) {
+    cls::rbd::MirrorSnapshotNamespace ns{
+      ((m_flags & CREATE_PRIMARY_FLAG_DEMOTED) != 0 ?
+        cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY_DEMOTED :
+        cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY),
+      m_mirror_peers_uuids[i], "", CEPH_NOSNAP};
+    ns.group_spec = m_image_ctxs[i]->group_spec;
+    ns.group_snap_id = m_group_snap_id;
+
+    ldout(m_cct, 15) << "name=" << m_snap_names[i] << ", "
+                     << "ns=" << ns << dendl;
```

Comment:
If you get a chance to respin this again, just consider this minor change:
```suggestion
    ldout(m_cct, 15) << "creating snapshot: image_id=" << m_image_ctxs[i]->id 
                     << ", name=" << m_snap_names[i] << ", "
                     << "ns=" << ns << dendl;
```